### PR TITLE
Ensure project can be published as an NPM module

### PR DESCRIPTION
**Pull request by dfunckt**
Change-type: patch

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,12 +2,8 @@ FROM balena/open-balena-base:v4.4.2
 
 EXPOSE 80
 
-COPY package.json package-lock.json /usr/src/app/
-RUN npm ci --unsafe-perm --production && npm cache clean --force
-
 COPY . /usr/src/app
-
-RUN npx tsc --noEmit
+RUN npm ci --unsafe-perm --production && npm cache clean --force
```
Comment (Page-) [inline Dockerfile]: Why did you remove the separate
```
COPY package.json package-lock.json /usr/src/app/
RUN npm ci --unsafe-perm --production && npm cache clean --force
```
?
The new version reduces the caching a lot so will slow down builds

File: package.json [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
   "repository": {
     "type": "git",
     "url": "https://github.com/balena-io/open-balena-api"
   },
+  "main": "dist/index",
+  "files": [
+    "dist/"
+  ],
+  "scripts": {
+    "build": "npm run clean && tsc --project . && copyup \"src/**/*.sbvr\" \"src/**/*.sql\" dist/",
```
Comment (Page-) [inline package.json]: Is `copyup` from this `copyfiles` dep?

File: package.json [hunk 3]
```diff
@@ -38,6 +48,7 @@
     "compression": "^1.7.3",
     "cookie-parser": "^1.4.1",
     "cookie-session": "^1.0.0",
+    "copyfiles": "^2.1.0",
```
Comment (Page-) [inline package.json]: This looks like it should be a dev dep

File: package.json [hunk 4]
```diff
@@ -60,13 +71,19 @@
     "request": "^2.85.0",
     "resin-device-config": "^4.2.0",
     "resin-semver": "^1.4.0",
+    "rimraf": "^2.6.1",
```
Comment (Page-) [inline package.json]: This also seems like a dev dep

File: src/hooks/index.ts [hunk 5]
```diff
@@ -13,8 +9,14 @@ sbvrUtils.addPureHook('all', 'all', 'all', {
 	},
 });
 
-readdirAsync(__dirname + '/resources').each(initScript => {
-	if (/\.ts$/.test(initScript)) {
-		return require(__dirname + '/resources/' + initScript);
-	}
-});
+import('./resources/api_key');
```
Comment (Page-) [inline src/hooks/index.ts]: Any reason for these being async imports?  It looks like they could just be imported directly since we're at the top level

File: src/lib/device-types/index.ts [hunk 6]
```diff
@@ -12,7 +12,7 @@ import {
 } from './build-info-facade';
 import { getImageKey, IMAGE_STORAGE_PREFIX, listFolders } from './storage';
 
-const { BadRequestError, NotFoundError } = sbvrUtils;
+export const { BadRequestError, NotFoundError } = sbvrUtils;
```
Comment (Page-) [inline src/lib/device-types/index.ts]: Why are these being exported now?

File: src/lib/errors.ts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import TypedError = require('typed-error');
 import { sbvrUtils } from '../platform';
 
-const { NotFoundError } = sbvrUtils;
+export const { NotFoundError } = sbvrUtils;
```
Comment (Page-) [inline src/lib/errors.ts]: Why is this being exported now?

File: src/routes/services.ts [hunk 8]
```diff
@@ -31,16 +32,19 @@ export const vpn = {
 				captureException(err, 'Error authenticating device for VPN', { req });
 				res.status(500).send(err);
 			}),
-	clientConnect: (req: Request, res: Response) => {
+	clientConnect: (req: Request, res: Response): Promise<void> => {
 		const body = req.body || {};
 		if (!body.common_name) {
-			return res.sendStatus(400);
+			res.sendStatus(400);
+			return null as any;
```
Comment (Page-) [inline src/routes/services.ts]: Why is this not just `return`?  That would avoid this odd type casting

File: src/routes/services.ts [hunk 9]
```diff
@@ -31,16 +32,19 @@ export const vpn = {
...
 		}
 		if (!body.virtual_address) {
-			return res.sendStatus(400);
+			res.sendStatus(400);
+			return null as any;
```
Comment (Page-) [inline src/routes/services.ts]: Why is this not just `return`?  That would avoid this odd type casting

File: src/routes/services.ts [hunk 10]
```diff
@@ -31,16 +32,19 @@ export const vpn = {
...
 		}
 		if (!body.service_id) {
-			return res.sendStatus(400);
+			res.sendStatus(400);
+			return null as any;
```
Comment (Page-) [inline src/routes/services.ts]: Why is this not just `return`?  That would avoid this odd type casting

File: src/routes/services.ts [hunk 11]
```diff
@@ -31,16 +32,19 @@ export const vpn = {
 				captureException(err, 'Error authenticating device for VPN', { req });
 				res.status(500).send(err);
 			}),
-	clientConnect: (req: Request, res: Response) => {
+	clientConnect: (req: Request, res: Response): Promise<void> => {
```
Comment (Page-) [inline src/routes/services.ts]: The real response type is `void | Promise<void>` - which is why I think you've done the weird type casting below

File: src/routes/services.ts [hunk 12]
```diff
@@ -67,10 +71,11 @@ export const vpn = {
 			});
 	},
 
-	clientDisconnect: (req: Request, res: Response) => {
+	clientDisconnect: (req: Request, res: Response): Promise<void> => {
```
Comment (Page-) [inline src/routes/services.ts]: The real response type is `void | Promise<void>` - which is why I think you've done the weird type casting below

File: Dockerfile (hunk 1 above)
Comment (dfunckt) [reply Dockerfile]: `npm` will invoke the `prepare` script during `npm ci` which triggers a build of the project. Without first copying the whole project into `/usr/src/app` there are no files to compile (or a tsconfig.json) and it fails. I don't like this TBH, because we're compiling sources during build but aren't really using them -- open-balena-api still starts up like resin-api, compiling on-demand. Any ideas?

File: package.json (hunk 2 above)
Comment (dfunckt) [reply package.json]: Yes, `copyup` is a "binary" equivalent to `copyfiles --up 1 ...`. I could change back to that if you think it's clearer.

File: package.json (hunk 3 above)
Comment (dfunckt) [reply package.json]: It can't unfortunately because we're triggering a build during `npm ci` when building the image.

File: Dockerfile (hunk 1 above)
Comment (Page-) [reply Dockerfile]: There's a `--ignore-scripts` for `npm i` that might work for ci?

File: package.json (hunk 2 above)
Comment (Page-) [reply package.json]: No it's fine, I just wasn't sure if it was some mac binary or not

File: Dockerfile (hunk 1 above)
Comment (dfunckt) [reply Dockerfile]: Ah, interesting, I'll take a look.

File: package.json (hunk 3 above)
Comment (Page-) [reply package.json]: `npm ci` will still install dev deps unless you specify `npm ci --production`

File: src/lib/device-types/index.ts (hunk 6 above)
Comment (dfunckt) [reply src/lib/device-types/index.ts]: TS complains we're extending a private type below.

File: src/lib/errors.ts (hunk 7 above)
Comment (dfunckt) [reply src/lib/errors.ts]: Same as above -- TS complains.

File: src/routes/services.ts (hunk 8 above)
Comment (dfunckt) [reply src/routes/services.ts]: TS complains again :) It treats `return` as equivalent to `return undefined` and complains the signature is incompatible.

File: src/lib/device-types/index.ts (hunk 6 above)
Comment (Page-) [reply src/lib/device-types/index.ts]: Ah right, because `declaration: true` was added, that kinda sucks because it's not actually private, just not exported by this module

File: src/routes/services.ts (hunk 8 above)
Comment (Page-) [reply src/routes/services.ts]: Well the signature *is* incompatible - it specifies a promise is returned when it's not

File: Dockerfile (hunk 1 above)
Comment (dfunckt) [reply Dockerfile]: That worked nicely 👍

File: Dockerfile [hunk 13]
```diff
@@ -3,7 +3,7 @@ FROM balena/open-balena-base:v4.4.2
 EXPOSE 80
 
 COPY package.json package-lock.json /usr/src/app/
-RUN npm ci --unsafe-perm --production && npm cache clean --force
+RUN npm ci --unsafe-perm --ignore-scripts && npm cache clean --force
```
Comment (Page-) [inline Dockerfile]: You probably still want the `--production`

Comment (dfunckt) [reply Dockerfile]: Oh right, we're not building the project anymore.

File: src/hooks/index.ts [hunk 14]
```diff
@@ -13,8 +9,14 @@ sbvrUtils.addPureHook('all', 'all', 'all', {
 	},
 });
 
-readdirAsync(__dirname + '/resources').each(initScript => {
-	if (/\.ts$/.test(initScript)) {
-		return require(__dirname + '/resources/' + initScript);
-	}
-});
+import * as _api_key from './resources/api_key';
```
Comment (Page-) [inline src/hooks/index.ts]: Just to check, these imports are still in the generated js right?

Comment (dfunckt) [reply src/hooks/index.ts]: They're not… good catch. Fixed.
